models/varnet/varnet.py

```diff
@@ -0,0 +1,356 @@
+        vxy = self.cov_norm * (uxy - ux * uy)
+        A1, A2, B1, B2 = (2 * ux * uy + C1, 2 * vxy + C2, ux ** 2 + uy ** 2 + C1, vx + vy + C2)
+        D = B1 * B2
+        S = (A1 * A2) / D
+        return 1 - S.mean()
+
+
+class NormUnet(nn.Module):
+    def __init__(self, chans, num_pools):
+        super().__init__()
+        self.unet = UnetModel(
+                in_chans=2,
+                out_chans=2,
+                chans=chans,
+                num_pool_layers=num_pools,
+                drop_prob=0
+            )
+
+    def complex_to_chan_dim(self, x):
+        b, c, h, w, two = x.shape
+        assert two == 2
+        return x.permute(0, 4, 1, 2, 3).contiguous().view(b, 2 * c, h, w)
+
+    def chan_complex_to_last_dim(self, x):
+        b, c2, h, w = x.shape
+        assert c2 % 2 == 0
+        c = c2 // 2
+        return x.view(b, 2, c, h, w).permute(0, 2, 3, 4, 1)
+
+    def norm(self, x):
+        # Group norm
+        b, c, h, w = x.shape
+        x = x.contiguous().view(b, 2, c // 2 * h * w)
+        mean = x.mean(dim=2).view(b, 2, 1, 1, 1).expand(b, 2, c // 2, 1, 1).contiguous().view(b, c, 1, 1)
+        std = x.std(dim=2).view(b, 2, 1, 1, 1).expand(b, 2, c // 2, 1, 1).contiguous().view(b, c, 1, 1)
+        x = x.view(b, c, h, w)
+        return (x - mean) / std, mean, std
+
+    def unnorm(self, x, mean, std):
+        return x * std + mean
+
+    def pad(self, x):
+        def floor_ceil(n):
+            return math.floor(n), math.ceil(n)
+
+        b, c, h, w = x.shape
+        w_mult = ((w - 1) | 15) + 1
+        h_mult = ((h - 1) | 15) + 1
+        w_pad = floor_ceil((w_mult - w) / 2)
+        h_pad = floor_ceil((h_mult - h) / 2)
+        x = F.pad(x, w_pad + h_pad)
+        return x, (h_pad, w_pad, h_mult, w_mult)
+
+    def unpad(self, x, h_pad, w_pad, h_mult, w_mult):
+        return x[..., h_pad[0]:h_mult - h_pad[1], w_pad[0]:w_mult - w_pad[1]]
+
+    def forward(self, x):
+        x = self.complex_to_chan_dim(x)
+        x, mean, std = self.norm(x)
+        x, pad_sizes = self.pad(x)
+        x = self.unet(x)
+        x = self.unpad(x, *pad_sizes)
+        x = self.unnorm(x, mean, std)
+        x = self.chan_complex_to_last_dim(x)
+        return x
+
+
+class VarNetBlock(nn.Module):
+    def __init__(self, model):
+        super(VarNetBlock, self).__init__()
+        self.model = model
+        self.dc_weight = nn.Parameter(torch.ones(1))
+        self.register_buffer('zero', torch.zeros(1, 1, 1, 1, 1))
+
+    def forward(self, current_kspace, ref_kspace, mask, sens_maps):
+        def sens_expand(x):
+            return T.fft2(T.complex_mul(x, sens_maps))
+
+        def sens_reduce(x):
+            x = T.ifft2(x)
+            return T.complex_mul(x, T.complex_conj(sens_maps)).sum(dim=1, keepdim=True)
+
+        def soft_dc(x, input):
+            return torch.where(mask, x - ref_kspace, self.zero) * self.dc_weight
+
+        return current_kspace - \
+                soft_dc(current_kspace, input) - \
```

Comment:
The variable input is not passed to the forward() of the VarNetBlock class. Am I missing something here?

Reply:
Yeah i'm not sure how this was running considering input wasn't defined (it seemed to work though). Either way I remove input from this function altogether.